benches/r1cs.rs

```diff
@@ -69,128 +69,111 @@ For K = 1:
-        }
-        let k = x.len();
-        if k == 1 {
-            cs.add_constraint([(x[0].0, -one), (y[0].0, one)].iter().collect());
-            return Ok(());
-        }
-        // Make last x multiplier for i = k-1 and k-2
-        let mut mulx_left = x[k - 1].1 + neg_z;
-        let mut mulx_right = x[k - 2].1 + neg_z;
-        let mut mulx_out = mulx_left * mulx_right;
-        let mut mulx_out_var_prev = KShuffleGadget::multiplier_helper(
-            cs,
-            neg_z,
-            mulx_left,
-            mulx_right,
-            mulx_out,
-            x[k - 1].0,
-            x[k - 2].0,
-            true,
-        )?;
-        // Make multipliers for x from i == [0, k-3]
-        for i in (0..k - 2).rev() {
-            mulx_left = mulx_out;
-            mulx_right = x[i].1 + neg_z;
-            mulx_out = mulx_left * mulx_right;
-            mulx_out_var_prev = KShuffleGadget::multiplier_helper(
-                cs,
-                neg_z,
-                mulx_left,
-                mulx_right,
-                mulx_out,
-                mulx_out_var_prev,
-                x[i].0,
-                false,
-            )?;
-        }
-        // Make last y multiplier for i = k-1 and k-2
-        let mut muly_left = y[k - 1].1 - z;
-        let mut muly_right = y[k - 2].1 - z;
-        let mut muly_out = muly_left * muly_right;
-        let mut muly_out_var_prev = KShuffleGadget::multiplier_helper(
-            cs,
-            neg_z,
-            muly_left,
-            muly_right,
-            muly_out,
-            y[k - 1].0,
-            y[k - 2].0,
-            true,
-        )?;
-        // Make multipliers for y from i == [0, k-3]
-        for i in (0..k - 2).rev() {
-            muly_left = muly_out;
-            muly_right = y[i].1 + neg_z;
-            muly_out = muly_left * muly_right;
-            muly_out_var_prev = KShuffleGadget::multiplier_helper(
-                cs,
-                neg_z,
-                muly_left,
-                muly_right,
-                muly_out,
-                muly_out_var_prev,
-                y[i].0,
-                false,
-            )?;
-        }
-        // Check equality between last x mul output and last y mul output
-        cs.add_constraint(
-            [(muly_out_var_prev, -one), (mulx_out_var_prev, one)]
-                .iter()
-                .collect(),
-        );
-        Ok(())
+/// Enforces that the output variables `y` are a valid reordering of the inputs variables `x`.
+/// The inputs and outputs are all tuples of the `Variable, Assignment`, where the `Assignment`
+/// can be either assigned as `Value::Scalar` or unassigned as `Missing`.
+pub fn fill_cs<CS: ConstraintSystem>(
+    cs: &mut CS,
+    x: Vec<(Variable, Assignment)>,
+    y: Vec<(Variable, Assignment)>,
+) -> Result<(), KShuffleError> {
+    let one = Scalar::one();
+    let z = cs.challenge_scalar(b"k-scalar shuffle challenge");
+    let neg_z = -z;
```

Comment:
nitpick: we can make the code a bit cleaner by just passing `z` and subtracting it when needed - the difference in performance of `Add` and `Sub` is really insignificant if there is any.